cmd/deploy/input.go

```diff
@@ -0,0 +1,246 @@
+	for d, ref := range normal.Lookup {
+		ref.Tag = ""
+		ref.Digest = ""
+		normal.Lookup[d] = ref
+	}
+
+	switch normal.Type {
+	case typeManifest:
+		if normal.CopyFrom == nil {
+			// instead of asking for mediaType explicitly, we'll enforce that any manifest we push *must* specify mediaType in the manifest itself (which then is *not* a restriction which applies to any children we copy); see https://github.com/opencontainers/distribution-spec/security/advisories/GHSA-mc8v-mgrf-8f4m and https://github.com/opencontainers/image-spec/security/advisories/GHSA-77vh-xpmg-72qh
+			var mediaTypeHaver struct {
+				// https://github.com/opencontainers/image-spec/blob/v1.1.0/specs-go/v1/index.go#L25
+				// https://github.com/opencontainers/image-spec/blob/v1.1.0/specs-go/v1/manifest.go#L24
+				MediaType string `json:"mediaType"`
+			}
+			if err := json.Unmarshal(normal.Data, &mediaTypeHaver); err != nil {
+				return normal, fmt.Errorf("%s: failed to parse %s data for mediaType: %w", debugId, normal.Type, err)
+			}
+			if mediaTypeHaver.MediaType == "" {
+				// we could just leave this blank and leave it up to the registry to reject instead: https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#push:~:text=Clients%20SHOULD%20set,the%20mediaType%20field.
+				// however, PushManifest expects mediaType: https://github.com/cue-labs/oci/blob/f3720d0e1bec6540a9b3c8783af010f51ad5cc95/ociregistry/ociclient/writer.go#L53
+				// and our logic for pushing children needs to know the mediaType (see the GHSAs referenced above)
+				return normal, fmt.Errorf("%s: pushing manifest but missing 'mediaType'", debugId)
+			}
+			normal.Do = func(ctx context.Context, dstRef registry.Reference) (ociregistry.Descriptor, error) {
+				return registry.EnsureManifest(ctx, dstRef, normal.Data, mediaTypeHaver.MediaType, normal.Lookup)
+			}
+		} else {
+			normal.Do = func(ctx context.Context, dstRef registry.Reference) (ociregistry.Descriptor, error) {
+				return registry.CopyManifest(ctx, *normal.CopyFrom, dstRef, normal.Lookup)
+			}
+		}
+
+	case typeBlob:
+		if normal.CopyFrom == nil {
+			normal.Do = func(ctx context.Context, dstRef registry.Reference) (ociregistry.Descriptor, error) {
+				return registry.EnsureBlob(ctx, dstRef, int64(len(normal.Data)), bytes.NewReader(normal.Data))
+			}
+		} else {
+			if normal.CopyFrom.Digest == "" {
+				return normal, fmt.Errorf("%s: blobs are always by-digest, and thus need a digest: %s", debugId, normal.CopyFrom)
+			}
+			normal.Do = func(ctx context.Context, dstRef registry.Reference) (ociregistry.Descriptor, error) {
+				return registry.CopyBlob(ctx, *normal.CopyFrom, dstRef)
+			}
+		}
+
+	default:
+		panic("unknown type: " + string(normal.Type))
+		// panic instead of error because this should've already been handled/normalized above (so this is a coding error, not a runtime error)
+	}
+
+	return normal, nil
```

Comment:
That is quite a function.

Reply:
Yeahhhhhhh -- I actually wrote this as part of `main()` itself, and when I stepped back from the implementation I almost died and thus forced myself to re-write it in such a way that I could load it up with unit tests so I could have higher confidence that all the edge cases in every branch of this behemoth are actually covered properly.